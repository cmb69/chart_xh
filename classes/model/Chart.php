<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Chart_XH.
 *
 * Chart_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Chart_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Chart_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Chart\Model;

use DOMDocument;
use DOMElement;
use DOMNode;
use Plib\Document2 as Document;
use Plib\DocumentStore2 as DocumentStore;

final class Chart implements Document
{
    public const TYPES = [
        "line",
        "bar",
        "horizontal-bar",
        "pie",
        "doughnut",
        "semi-pie",
        "semi-doughnut",
        "polar-area",
    ];

    private string $name;
    private string $caption;
    private string $type;
    private bool $transposed;
    private string $aspectRatio;
    /** @var list<string> */
    private array $labels = [];
    /** @var list<Dataset> */
    private array $datasets = [];

    public static function new(string $key): self
    {
        return new self(basename($key, ".xml"), "", "line", false, "3/2");
    }

    public static function fromString(string $contents, string $key): ?self
    {
        if ($contents === "") {
            return null;
        }
        $doc = new DOMDocument("1.0", "UTF-8");
        if (!$doc->loadXML($contents)) {
            return null;
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return null;
        }
        assert($doc->documentElement instanceof DOMElement);
        $chart = $doc->documentElement;
        $that = new self(
            basename($key, ".xml"),
            $chart->getAttribute("caption"),
            $chart->getAttribute("type"),
            (bool) $chart->getAttribute("transposed"),
            $chart->getAttribute("aspect-ratio")
        );
        foreach ($chart->childNodes as $childNode) {
            assert($childNode instanceof DOMNode);
            if ($childNode->nodeName === "label") {
                assert($childNode instanceof DOMElement);
                $that->labels[] = (string) $childNode->nodeValue;
            } elseif ($childNode->nodeName === "dataset") {
                assert($childNode instanceof DOMElement);
                $that->datasets[] = Dataset::fromXml($childNode);
            }
        }
        return $that;
    }

    public static function create(string $name, DocumentStore $store): self
    {
        $that = $store->create("$name.xml", self::class);
        assert($that instanceof self);
        return $that;
    }

    public static function read(string $name, DocumentStore $store): ?self
    {
        return $store->read("$name.xml", self::class);
    }

    public static function update(string $name, DocumentStore $store): ?self
    {
        return $store->update("$name.xml", self::class);
    }

    public function __construct(string $name, string $caption, string $type, bool $transposed, string $aspectRatio)
    {
        $this->name = $name;
        $this->caption = $caption;
        $this->type = $type;
        $this->transposed = $transposed;
        $this->aspectRatio = $aspectRatio;
    }

    public function name(): string
    {
        return $this->name;
    }

    public function caption(): string
    {
        return $this->caption;
    }

    public function type(): string
    {
        return $this->type;
    }

    public function transposed(): bool
    {
        return $this->transposed;
    }

    public function aspectRatio(): string
    {
        return $this->aspectRatio;
    }

    /** @return list<string> */
    public function labels(): array
    {
        return $this->labels;
    }

    /** @return list<Dataset> */
    public function datasets(): array
    {
        return $this->datasets;
    }

    public function setCaption(string $caption): void
    {
        $this->caption = $caption;
    }

    public function setType(string $type): void
    {
        $this->type = $type;
    }

    public function setTransposed(bool $transposed): void
    {
        $this->transposed = $transposed;
    }

    public function setAspectRatio(string $aspectRatio): void
    {
        $this->aspectRatio = $aspectRatio;
    }

    public function purgeLabels(): void
    {
        $this->labels = [];
    }

    public function addLabel(string $label): void
    {
        $this->labels[] = $label;
    }

    public function purgeDatasets(): void
    {
        $this->datasets = [];
    }

    public function addDataset(string $label, string $color): Dataset
    {
        return $this->datasets[] = new Dataset($label, $color);
    }

    public function toString(): string
    {
        $doc = new DOMDocument('1.0', 'UTF-8');
        $chart = $doc->createElement('chart');
        $chart->setAttribute("caption", $this->caption);
        $chart->setAttribute("type", $this->type);
        $chart->setAttribute("transposed", (string) $this->transposed);
        $chart->setAttribute("aspect-ratio", $this->aspectRatio);
        $doc->appendChild($chart);
        foreach ($this->labels as $label) {
            $elt = $doc->createElement("label");
            $elt->nodeValue = $label;
            $chart->appendChild($elt);
        }
        foreach ($this->datasets as $dataset) {
            $chart->appendChild($dataset->toXml($doc));
        }
        if (!$doc->relaxNGValidate(__DIR__ . "/../../chart.rng")) {
            return "";
        }
        $doc->formatOutput = true;
        return (string) $doc->saveXML();
    }
}
