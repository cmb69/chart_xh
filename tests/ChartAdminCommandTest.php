<?php

namespace Chart;

use ApprovalTests\Approvals;
use Chart\Model\Chart;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\CsrfProtector;
use Plib\DocumentStore2 as DocumentStore;
use Plib\FakeRequest;
use Plib\View;

class ChartAdminCommandTest extends TestCase
{
    private DocumentStore $store;
    /** @var CsrfProtector&Stub */
    private $csrfProtector;
    private View $view;

    public function setUp(): void
    {
        vfsStream::setup("root");
        $this->store = new DocumentStore(vfsStream::url("root/"));
        $chart = Chart::create("test", $this->store);
        $dataset = $chart->addDataset("one", "#ff0000");
        $dataset->addValue(1);
        $dataset->addValue(2);
        $dataset->addValue(3);
        $this->store->commit();
        $this->csrfProtector = $this->createStub(CsrfProtector::class);
        $this->csrfProtector->method("token")->willReturn("0123456789ABCDEF");
        $this->view = new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["chart"]);
    }

    private function sut(): ChartAdminCommand
    {
        return new ChartAdminCommand("./", $this->store, $this->csrfProtector, $this->view);
    }

    public function testRendersOverview(): void
    {
        $request = new FakeRequest();
        $response = $this->sut()($request);
        $this->assertSame("Chart – Administration", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testRendersEditorForNewChart(): void
    {
        $request = new FakeRequest(["url" => "http://example.com/?&chart&admin=plugin_main&action=create"]);
        $response = $this->sut()($request);
        $this->assertSame("Chart – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testCreatesNewChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertFileExists(vfsStream::url("root/new.xml"));
        $this->assertSame("http://example.com/?&chart&admin=plugin_main", $response->location());
    }

    public function testCreatingIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToSaveNewChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=create",
            "post" => [
                "chart_do" => "",
                "name" => "new",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the chart “new”!", $response->output());
    }

    public function testRendersEditorForUpdate(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
        ]);
        $response = $this->sut()($request);
        $this->assertSame("Chart – Edit", $response->title());
        Approvals::verifyHtml($response->output());
    }

    public function testsReportsThatNoChartIsSelected(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You did not select a chart!", $response->output());
    }

    public function testReportsFailureToLoadChart(): void
    {
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=does-not-exist",
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the chart “does-not-exist”!", $response->output());
    }

    public function testUpdatesChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "new chart caption",
                "labels" => "one,two,three",
                "datasets" => '[{"label":"one","color":"#ff0000","values":[1,2,3]}]'
            ],
        ]);
        $response = $this->sut()($request);
        $chart = Chart::read("test", $this->store);
        $this->assertSame("new chart caption", $chart->caption());
        $this->assertEquals(["one", "two", "three"], $chart->labels());
        $this->assertCount(1, $chart->datasets());
        $dataset = $chart->datasets()[0];
        $this->assertEquals("one", $dataset->label());
        $this->assertEquals("#ff0000", $dataset->color());
        $this->assertCount(3, $dataset->values());
        $this->assertSame("http://example.com/?&chart&admin=plugin_main&chart_name=test", $response->location());
    }

    public function testsReportsThatNoChartIsSelectedWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update",
            "post" => [
                "chart_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You did not select a chart!", $response->output());
    }

    public function testsReportsFailureToLoadChartWhenUpdating(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=does-not-exist",
            "post" => [
                "chart_do" => "",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot load the chart “does-not-exist”!", $response->output());
    }

    public function testUpdateIsCsrfProtected(): void
    {
        $this->csrfProtector->method("check")->willReturn(false);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "a new caption",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("You are not authorized to conduct this action!", $response->output());
    }

    public function testReportsFailureToUpdateChart(): void
    {
        $this->csrfProtector->method("check")->willReturn(true);
        vfsStream::setQuota(0);
        $request = new FakeRequest([
            "url" => "http://example.com/?&chart&admin=plugin_main&action=update&chart_name=test",
            "post" => [
                "chart_do" => "",
                "caption" => "a new caption",
            ],
        ]);
        $response = $this->sut()($request);
        $this->assertStringContainsString("Cannot save the chart “test”!", $response->output());
    }
}
