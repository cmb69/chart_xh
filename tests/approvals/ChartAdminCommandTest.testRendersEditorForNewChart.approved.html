
<script type="module" src="/admin.js?&amp;v=1.0-dev"></script>
<article class="chart_edit">
  <h1>Chart â€“ Edit</h1>
  <form method="post">
    <p>
      <label>
        <span>Name</span>
        <span class="chart_help">(only a-z, 0-9, and hyhpens)</span>
        <input name="name" value=""  required pattern="[a-z0-9\-]+">
      </label>
    </p>
    <p>
      <label>
        <span>Caption</span>
        <span class="chart_help">(HTML)</span>
        <textarea name="caption"></textarea>
      </label>
    </p>
    <p>
      <label>
        <span>Type</span>
        <select name="type">
          <option value="line" selected>Line</option>
          <option value="bar" >Bar</option>
          <option value="horizontal-bar" >Bar (horizontal)</option>
          <option value="pie" >Pie</option>
          <option value="doughnut" >Doughnut</option>
          <option value="semi-pie" >Semi-Pie</option>
          <option value="semi-doughnut" >Semi-Doughnut</option>
          <option value="polar-area" >Polar Area</option>
        </select>
      </label>
      <label>
        <input type="checkbox" name="transposed" >
        <span>transposed</span>
      </label>
    </p>
    <p>
      <label>
        <span>Aspect Ratio</span>
        <span class="maps_help">(with/height)</span>
        <input name="aspect_ratio" value="3/2" required pattern="\d+/\d+">
    <p class="chart_labels">
      <label>
        <span>Labels</span>
        <span class="chart_help">(comma separated strings)</span>
        <textarea name="labels"></textarea>
      </label>
    </p>
    <p class="chart_datasets">
      <label>
        <span>Datasets</span>
        <textarea name="datasets">[]</textarea>
      </label>
    </p>
    <table style="display: none">
      <caption>Datasets</caption>
      <colgroup>
        <col>
        <col>
        <col>
        <col>
      </colgroup>
      <thead>
        <tr>
          <th>Label</th>
          <th>Color</th>
          <th>Values <span class="chart_help">(comma separated numbers; decimal separator is dot)</span></th>
          <th><button type="button" class="chart_add_dataset"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" width="1em" height="1em" fill="currentColor"><title>Add Dataset</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M64 32C28.7 32 0 60.7 0 96L0 416c0 35.3 28.7 64 64 64l320 0c35.3 0 64-28.7 64-64l0-320c0-35.3-28.7-64-64-64L64 32zM200 344l0-64-64 0c-13.3 0-24-10.7-24-24s10.7-24 24-24l64 0 0-64c0-13.3 10.7-24 24-24s24 10.7 24 24l0 64 64 0c13.3 0 24 10.7 24 24s-10.7 24-24 24l-64 0 0 64c0 13.3-10.7 24-24 24s-24-10.7-24-24z"/></svg></button></th>
        </tr>
      </thead>
      <tbody>
      </tbody>
      <tfoot style="display: none">
        <tr>
          <td><input></td>
          <td><input type="color"></td>
          <td>
            <textarea class="chart_values"></textarea>
            <button type="button" class="chart_edit_values"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" width="1em" height="1em" fill="currentColor"><title>Edit Values</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M410.3 231l11.3-11.3-33.9-33.9-62.1-62.1L291.7 89.8l-11.3 11.3-22.6 22.6L58.6 322.9c-10.4 10.4-18 23.3-22.2 37.4L1 480.7c-2.5 8.4-.2 17.5 6.1 23.7s15.3 8.5 23.7 6.1l120.3-35.4c14.1-4.2 27-11.8 37.4-22.2L387.7 253.7 410.3 231zM160 399.4l-9.1 22.7c-4 3.1-8.5 5.4-13.3 6.9L59.4 452l23-78.1c1.4-4.9 3.8-9.4 6.9-13.3l22.7-9.1 0 32c0 8.8 7.2 16 16 16l32 0zM362.7 18.7L348.3 33.2 325.7 55.8 314.3 67.1l33.9 33.9 62.1 62.1 33.9 33.9 11.3-11.3 22.6-22.6 14.5-14.5c25-25 25-65.5 0-90.5L453.3 18.7c-25-25-65.5-25-90.5 0zm-47.4 168l-144 144c-6.2 6.2-16.4 6.2-22.6 0s-6.2-16.4 0-22.6l144-144c6.2-6.2 16.4-6.2 22.6 0s6.2 16.4 0 22.6z"/></svg></button>
            <button type="button" class="chart_apply_values" style="display: none"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512" width="1em" height="1em" fill="currentColor"><title>Apply Values</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M438.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L160 338.7 393.4 105.4c12.5-12.5 32.8-12.5 45.3 0z"/></svg></button>
          </td>
          <td>
            <button type="button" class="chart_move_dataset"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512" width="1em" height="1em" fill="currentColor"><title>Move Dataset</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M350 177.5c3.8-8.8 2-19-4.6-26l-136-144C204.9 2.7 198.6 0 192 0s-12.9 2.7-17.4 7.5l-136 144c-6.6 7-8.4 17.2-4.6 26s12.5 14.5 22 14.5l88 0 0 192c0 17.7-14.3 32-32 32l-80 0c-17.7 0-32 14.3-32 32l0 32c0 17.7 14.3 32 32 32l80 0c70.7 0 128-57.3 128-128l0-192 88 0c9.6 0 18.2-5.7 22-14.5z"/></svg></button>
            <button type="button" class="chart_delete_dataset"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512" width="1em" height="1em" fill="currentColor"><title>Delete Dataset</title><!--!Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free Copyright 2025 Fonticons, Inc.--><path d="M135.2 17.7L128 32 32 32C14.3 32 0 46.3 0 64S14.3 96 32 96l384 0c17.7 0 32-14.3 32-32s-14.3-32-32-32l-96 0-7.2-14.3C307.4 6.8 296.3 0 284.2 0L163.8 0c-12.1 0-23.2 6.8-28.6 17.7zM416 128L32 128 53.2 467c1.6 25.3 22.6 45 47.9 45l245.8 0c25.3 0 46.3-19.7 47.9-45L416 128z"/></svg></button>
          </td>
        </tr>
      </tfoot>
    </table>
    <p class="chart_controls">
      <button name="chart_do">Save</button>
    </p>
    <input type="hidden" name="chart_token" value="0123456789ABCDEF">
  </form>
</article>
